core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -380,6 +380,14 @@ public CompletableFuture<Void> maybeInitialize() {
 
             // Update state to initializing to avoid any concurrent requests to be processed.
             partitionState = SharePartitionState.INITIALIZING;
+        } catch (Exception e) {
+            log.error("Failed to initialize the share partition: {}-{}", groupId, topicIdPartition, e);
+            completeInitializationWithException(future, e);
+            return future;
+        } finally {
+            lock.writeLock().unlock();
+        }
+        try {
```

Comment:
Shouldn't this `try` be inside of the response from persister read state, why to have a lock while invoking the persister read call?

Reply:
We need this `try` in case we have persister.readState throw  any exceptions. We even have a test case `testMaybeInitializeWithReadException` in `SharePartitionTest` that verifies this behaviour and it will start failing in case we get rid of this `try catch` block.

Reply:
> Shouldn't this try be inside of the response from persister read state, why to have a lock while invoking the persister read call?

The `try` block is not related to the lock; instead, it handles exceptions thrown by `persister.readState`. This introduces another inconsistency in error handling - The `persister.writeState` can also throw exceptions, but we do not use `try-catch` blocks to complete the `future` with an exception.

Reply:
We should follow up by unifying the error handling for all `persister` methods. Since `persister` uses `CompletableFuture` as returned type, we can leverage it to propagate exceptions instead of handling them separately in both `try-catch` and `whenComplete` blocks.

Reply:
@chia7712 , actually exceptions are getting handled when `persister.writeState` throws exceptions. We have a test case `testWriteShareGroupStateWithWriteException` in `SharePartitionTest` that verifies this beahviour.

Reply:
> actually exceptions are getting handled when persister.writeState throws exceptions. We have a test case testWriteShareGroupStateWithWriteException in SharePartitionTest that verifies this beahviour.

`testWriteShareGroupStateWithWriteException` verify the exception wrapped by `CompletableFuture`. However, `persister.writeState` could throw exception directly. The exception is not wrapped by `CompletableFuture` 

https://github.com/apache/kafka/blob/d00f0ecf1a1a082c97564f4b807e7a342472b57a/share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java#L82

Reply:
>We need this try in case we have persister.readState throw any exceptions

Hmmm, I was more in line with to have the try block when lock is required and not have `persister.readState/persister.writeState` in the try block as persister should return the completable future with exception, if failed, and should never throw an exception directly. The javadoc of persister do not specify throwing any exceptions directly.

>We even have a test case testMaybeInitializeWithReadException in SharePartitionTest that verifies this behaviour

We can change the expectation of the test in the 2nd case of the test. Also we should see the exception is being handled at top level if this happens but we can't write entire code in try/catch.

Reply:
@chia7712, agreed, my latest commit as mentioned on comment https://github.com/apache/kafka/pull/17957#discussion_r1866535510 contains the change where exceptions thrown directly are also handled now.

Reply:
> as persister should return the completable future with exception, if failed, and should never throw an exception directly. 

yes I love this style too :)

Reply:
@adixitconfluent I see your last update use the `try-catch` to handle exception thrown by `persister.readState/persister.writeState`. As @apoorvmittal10 comment and my previous comment (https://github.com/apache/kafka/pull/17957#discussion_r1866507046), maybe `persister` can leverage `CompletableFuture` to propagate the exception? WDYT?

Reply:
@chia7712, I can remove the `try catch` from `persister.readState/persister.writeState` calls and modify the test cases for the corresponding behaviour. From what I see in the persister code lines [here](https://github.com/apache/kafka/blob/trunk/share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java#L208) and [here](https://github.com/apache/kafka/blob/trunk/share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java#L118), `CompletableFuture` is being utilized to propagate the exception. I think the `try catch` in `persister.readState` was added as an additional check in the initial iterations of development, but we don't seem to need it now. If we're all aligned, then  I'll make this change and re-raise a review.

Reply:
> I think the try catch in persister.readState was added as an additional check in the initial iterations of development

maybe we can move the "additional check" into `persister.readState/persister.writeState`? That can ensure they don't throw exception directly. For example: `validate(request)` could cause exception. 

> we don't seem to need it now. If we're all aligned, then I'll make this change and re-raise a review.

yes, we don't need the `try-catch` if `persister.readState/persister.writeState` always propagate exception by `CompletableFuture`

Reply:
@chia7712 @apoorvmittal10 , I've moved the `try-catch` check  from `persister.readState/persister.writeState` to `DefaultStatePersister readState/writeState` methods. Had a discussion with @smjn regarding the scope of try catch block within `readState/writeState` methods, `validate(request)` could only throw such exception. Else everything is handled by the future.